Camera platform for animated Australian BOM radar imagery

maddenp opened the pull request:
## Description:

This PR adds a camera platform that delivers animated GIF loops of weather-radar imagery from the Australian Bureau of Meteorology.

The `bomradarcam.py` source file contains a thorough description of configuration options in its module docstring. Fuller external documentation will follow if this submission is accepted.

## Example entry for `configuration.yaml` (if applicable):

```yaml
camera:
  - platform: bomradarcam
    location: Sydney
```
## Checklist:
  - [x] The code change is tested and works locally.
  - [see below] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

I cannot get `tox` to work on my system but, following https://developers.home-assistant.io/docs/en/development_testing.html#testing-outside-of-tox, the code passed individual `flake8`, `pylint`, and `pydocstyle` tests. There are no unit tests at the moment.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.
(i.e. the code _does_ interact with devices, web services in this case)

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

Commit:
Add bomradarloop.py

Commit:
Delinting

Commit:
Delinting

Commit:
Delinting

Commit:
Work on logging statements

Commit:
Replace exception handling with if/else

Commit:
Fix exceptions

Commit:
Placate flake8

Commit:
Pylint delinting

Commit:
Update requirements_all.txt

Commit:
Add bomradarloop.py to .coveragerc

homeassistant commented:
Hi @maddenp, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#21183).

Once you do that we will be able to review and accept this pull request.

Thanks!

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+"""
+Provie animtaed GIF loops of BOM radar imagery.
```
Spelling errors

maddenp replied:
Thanks. Fixed, along with some other copy-editing in comments.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+"""
+Provie animtaed GIF loops of BOM radar imagery.
+
+This component provides animated GIF images of weather-radar imagery derived
```
This should be in the documentation.

maddenp replied:
Would you prefer to have it removed from the source file now, or simply _moved_ to the external documentation when that's written?

maddenp replied:
I've gone ahead and removed the extra docs and am now working toward a PR in home-assistant.io that includes them.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+
+
+def _validate_schema(cfg):
+    msg0 = "Specify either 'id' or 'location', not both"
```
Make them global or have them directly in the Invalid

maddenp replied:
Done.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+    return 'http://www.bom.gov.au/{}'.format(path)
+
+
+def _log(*msg):
```
Unnecessary, just call LOGGER.info when relevant

maddenp replied:
Done.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+
+        super().__init__()
+
+        self.hass = hass
```
This is done by entity platform, no need to do manually

maddenp replied:
Removed.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+
+    def get_background(self):
+        """
+        Get the background map image.
```
No newline before first line of text

maddenp replied:
Updated.

Kane610 commented on homeassistant/components/camera/bomradarloop.py:
```diff
@@ -0,0 +1,381 @@
+    name = config.get(CONF_NAME) or "BOM Radar Loop - {}".format(location)
+    outfn = config.get(CONF_OUTFN)
+    bomradarloop = BOMRadarLoop(
+        hass,
```
fewer lines, longer lines

maddenp replied:
Made fewer, but longer, lines.

Kane610 reviewed (changes requested):
Well written overall 👍

Commit:
Copy editing

Commit:
Inline error messages per PR request

Commit:
Eliminate logging functions

Commit:
Remove 'self.hass = hass' statement

Commit:
No newline before """ in docstrings

Commit:
Fewer, but longer, lines per PR requerst

Commit:
Moving material from module docstring to web docs

Commit:
Update Townsville rada default values

maddenp commented:
Documentation PR: https://github.com/home-assistant/home-assistant.io/pull/8632

fabaff reviewed (changes requested):
All code that is dealing with the gif should be moved to a Python module that is published on PyPI.

Commit:
Refactor to use PyPI package

maddenp commented:
Refactored to use PyPI package. Updating home-assistant.io docs PR now.

Commit:
Update requirements_all.txt

maddenp commented:
I updated the platform name to avoid collision with the name of the PyPI package. I updated the description, but am not sure how to update the label. In any case, I think I've done what I can with this for now. Thanks for having another look when you have a chance.

maddenp commented:
It crossed my mind to check in on this and I see that a docs-missing label was added. Is that something I should be taking action on? If so, what does it mean? There's a PR for documentation (https://github.com/home-assistant/home-assistant.io/pull/8632#pullrequestreview-205212540), but maybe that's not what's meant.

balloob committed:
Bumped version to 0.91.0b0

balloob committed:
Fix YAML

balloob committed:
Bumped version to 0.91.0b1

awarecan committed:
Fix dev branch (#22493)

awarecan committed:
A very basic Circleci setup (#22503)

* Add circleci support

* Add buildpack-deps

* Install libudev-dev

* sudo

* always run test

* Add test report

* no sugar

* quite pytest

* better junit test result

* Add $CODE_COVERAGE env var

awarecan committed:
Fix Circleci config (#22509)

* Add libav depends on circleci

* tweak circleci config

awarecan committed:
Remove botocore dependency from credstash script (#22511)

* Remove botocore dependency from credstash script

* Update requirements_all.txt

* Update pylintrc

* Update credstash.py

awarecan committed:
Fix lint on dev (#22512)

## Description:

Fix a lint issue in credstash script.


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

mvn23 committed:
Update pyotgw to 0.4b3 (#22496)

dmulcahey committed:
Do not use zha default light polling (#22513)

* don't use default light polling

* review comment

zewelor committed:
Fixes for yeelight availbility state (#22502)

awarecan committed:
Fix tts Great Migration issue (#22539)

ktnrg45 committed:
PS4 bump to 0.5.2 (#22523)

* Bump pyps4 to 0.5.2

* Bump pyps4 to 0.5.2

* Bump pyps4 to 0.5.2

dmulcahey committed:
clean up channel configuration (#22534)

syssi committed:
Fix regression of the xiaomi_aqara config validation (#22435)

* Fix regression of the xiaomi_aqara config validation

* Make the key optional again

* Add base schema

* Remove the GW_MAC default

StevenLooman committed:
Upgrade to async_upnp_client==0.14.7 (#22543)

damarco committed:
Bump zigpy (#22545)

balloob committed:
Bumped version to 0.91.0b2

awarecan committed:
Set up Circleci workflow (#22519)

* Set up Circleci workflow

* Update python tag

* Add pre-test job to cache the requirements

* Upgrade pip itself

* Use 3.7 for lint

* Parallelize pylint

* Tweak run gen_requirements_all

* tweak cache key

balloob committed:
Updated frontend to 20190329.0

balloob committed:
Add support for streaming to ffmpeg (#22549)

balloob committed:
Fix platform warnings (#22551)

balloob committed:
Bumped version to 0.91.0b3

giefca committed:
Google assistant: add blinds trait for covers (#22336)

* Update const.py

* Update smart_home.py

* Update trait.py

* Update test_trait.py

* Update smart_home.py

* Update test_trait.py

* Update trait.py

* Update trait.py

* Update test_trait.py

* Update test_trait.py

* Update __init__.py

* Update test_trait.py

* Change email

* Trying to correct CLA

* Update __init__.py

* Update trait.py

* Update trait.py

* Update trait.py

* Update trait.py

* Update __init__.py

* Update test_trait.py

* Update test_google_assistant.py

* Update trait.py

* Update trait.py

* Update test_trait.py

* Update test_trait.py

drjared88 committed:
Update Amcrest component to SUPPORT_STREAM (#22553)

* Update camera.py

Update Amcrest component to SUPPORT_STREAM to allow streaming in the UI and Google Assistant.

* Update camera.py

awarecan committed:
Fix name conflict in tests (#22556)

* Fix name conflict in tests

* Lint

* Lint

CWhits committed:
Update Foscam component to support stream source (#22568)

* Update Foscam to support stream source

* Removing spaces and tabs

* Changing to Python3-style string formatting

* Adding '_media_port' to hopefully cover other models

drjared88 committed:
Update ONVIF component to SUPPORT_STREAM (#22569)

* Update Onvif component to SUPPORT_STREAM

* Update camera.py

* Update camera.py

* Update camera.py

Remove extra spaces.

* lookup URL when camera is added to hass and add extra guards

balloob committed:
Add stream to the default config (#22602)

balloob committed:
Updated frontend to 20190331.0

balloob committed:
Bumped version to 0.91.0b4

pvizeli committed:
Hass.io ingress (#22505)

* Fix API stream of snapshot / Add ingress

* fix lint

* Fix stream handling

* Cleanup api handling

* fix typing

* Set proxy header

* Use header constant

* Enable the ingress setup

* fix lint

* Fix name

* Fix tests

* fix lint

* forward params

* Add tests for ingress

* Cleanup cookie handling with aiohttp 3.5

* Add more tests

* Fix tests

* Fix lint

* Fix header handling for steam

* forward header too

* fix lint

* fix flake

awarecan committed:
Config CircleCI workflow (#22590)

* Add mypyrc to control typing check, add mypy to circle

* Add translation upload circlci job

pvizeli committed:
Update face_recognition to 1.2.3 (#22622)

balloob committed:
Disable Z-Wave autoheal (#22628)

Danielhiversen committed:
Handle disonnect bug in Tibber library (#22629)

pvizeli committed:
Support GET params for websocket ingress path (#22638)

awarecan committed:
Require static-check success first for rest of workflow (#22635)

* Require static-check success first

* Update config.yml

awarecan committed:
Add trusted networks deprecating warning (#22487)

* Add trusted networks deprecating warning

* Update auth.py

* Update auth.py

* Update auth.py

* Update auth.py

* Tweak

balloob committed:
Run PyLint under Python 3.5 (#22642)

* Run PyLint under Python 3.5

* Remove -q from pip install to debug

* Upgrade setuptools before install

* Use correct cache key for pylint

Jc2k committed:
Fix racy homekit_controller platform setup caused by #22368 (#22655)

CWhits committed:
Return 0 for failed Foscam streams (#22651)

* Update Foscam to support stream source

* Removing spaces and tabs

* Changing to Python3-style string formatting

* Adding '_media_port' to hopefully cover other models

* changing logic for success and return none

balloob committed:
Bumped version to 0.91.0b5

mvn23 committed:
Don't use room setpoint override in climate.opentherm_gw (#22656)

* Dont use DATA_ROOM_SETPOINT_OVRD in climate.opentherm_gw as it is unreliable with some thermostats.

* Show new target temperature immediately until the backend notices a change

* Only update target temp on the gateway if the value differs from the current target_temperature.

pvizeli committed:
Hass.io discovery flow deconz (#22623)

* Add Hass.io deCONZ discovery flow

* add bridge ID

* fix attribute

* fix strings

* Address comments

* Add test

* Add only instance / changed maybe later

awarecan committed:
Fix trusted networks auth provider warning message (#22671)

* Fix trusted networks auth provider warning message

* Update auth.py

dgomes committed:
Fix #22648 - Utility_meter would try to cancel a non existing task (#22669)

* don't cancel tariff that are paused

* test tariffs

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+Provide animated GIF loops of BOM radar imagery.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/bomradarcam/
```
We don't include the url in the docstring anymore. Just keep the very first line.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/bomradarcam/
+"""
+
```
Please remove this blank line.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+CONF_FRAMES = 'frames'
+CONF_ID = 'id'
+CONF_LOC = 'location'
+CONF_NAME = 'name'
```
Import this from const.py.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+CONF_DELTA = 'delta'
+CONF_FRAMES = 'frames'
+CONF_ID = 'id'
+CONF_LOC = 'location'
```
```py
CONF_LOCATION = 'location'
```

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+
+CONF_DELTA = 'delta'
+CONF_FRAMES = 'frames'
+CONF_ID = 'id'
```
Import this from const.py.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def _validate_schema(config):
+    if config.get('location'):
```
Use constant `CONF_LOCATION`.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+
+def _validate_schema(config):
+    if config.get('location'):
+        if config.get('id'):
```
Use `CONF_ID`.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+def _validate_schema(config):
+    if config.get('location'):
+        if config.get('id'):
+            raise vol.Invalid("Specify either 'id' or 'location', not both")
```
Instead use `vol.Exclusive` in the config schema.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+    else:
+        reqs = (config.get('id'), config.get('delta'), config.get('frames'))
+        if not all(reqs):
+            raise vol.Invalid("Specify 'id', 'delta' and 'frames' when"
```
Instead use `vol.Inclusive` in the schema.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+}), _validate_schema)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Rename `add_devices` to `add_entities`.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+class BOMRadarCam(Camera):
+    """A camera component producing animated BOM radar-imagery GIFs."""
+
+    def __init__(self, hass, name, location, radar_id, delta, frames, outfile):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+        super().__init__()
+        self._name = name
+        self._cam = BOMRadarLoop(location, radar_id, delta, frames, outfile,
+                                 LOGGER)
```
Don't pass the logger to the library. The library should use its own logger, one per module. This is the recommended way of implementing loggers according to standard library.

https://docs.python.org/3/library/logging.html#logger-objects

MartinHjelmare commented on homeassistant/components/camera/bomradarcam.py:
```diff
@@ -0,0 +1,75 @@
+    vol.Optional(CONF_OUTFILE): cv.string,
+    vol.Optional(CONF_FRAMES): cv.positive_int,
+    vol.Optional(CONF_ID): cv.string,
+    vol.Optional(CONF_LOC): cv.string,
```
Is it possible to validate what locations or ids are allowed?

What happens if a user configures an incorrect location?

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -219,6 +219,9 @@ blockchain==1.4.4
 # homeassistant.components.sensor.bme680
 # bme680==1.0.5
 
+# homeassistant.components.camera.bomradarcam
```
Move, `git mv`, the camera platform under a new bomradarcam component package. That's our new way of structuring modules.
```
homeassistant/components/bomradarcam/__init__.py
homeassistant/components/bomradarcam/camera.py
```

Make sure to update `.coveragerc` and run `script/gen_requirements_all.py` after the move.

pvizeli committed:
Fix ffmpeg default extra options (#22682)

balloob committed:
Deal with cover assumed state (#22673)

* Deal with cover assumed state

* Add docs

robbiet480 committed:
Add mobile_app notify platform (#22580)

* Add mobile_app notify platform

* Requested changes

* Fix incorrect param for status code

* Move push_registrations to notify platform file

* Trim down registration information sent in push

* quotes

* Use async version of load_platform

* Add warning for duplicate device names

* Switch to async_get_service

* add mobile_app.notify test

* Update tests/components/mobile_app/test_notify.py

* Update tests/components/mobile_app/test_notify.py

balloob committed:
Bumped version to 0.91.0

balloob committed:
Merge pull request #22688 from home-assistant/rc

0.91.0

Commit:
Merge branch 'master' into bomradarloop
